instaclustr/resource_kafka_topic.go

```diff
@@ -0,0 +1,419 @@
+	}
+	var topicExist bool
+	for _, t := range topicList.Topics {
+		if t == topic {
+			topicExist = true
+		}
+	}
+	if !topicExist {
+		d.SetId("")
+		log.Printf("[INFO] Topic %s not found in cluster %s.", topic, cluster_id)
+		return nil
+	}
+
+	// Cluster has to reach running state first
+	cluster, err := client.ReadCluster(cluster_id)
+	if err != nil {
+		return fmt.Errorf("[Error] Error in getting the status of the cluster: %w", err)
+	}
+	if cluster.ClusterStatus != "RUNNING" {
+		return fmt.Errorf("[Error] Cluster %s is not RUNNING. Currently in %s state", cluster_id, cluster.ClusterStatus)
+	}
+
+	log.Printf("[INFO] Reading the replication facotr and number of partitions of topic %s.", topic)
+	kafkaTopic, err := client.ReadKafkaTopic(cluster_id, topic)
+	if err != nil {
+		return fmt.Errorf("[Error] Error reading Kafka topic %s's replication facotr and number of partitions: %w", topic, err)
+	}
+	d.Set("replication_factor", kafkaTopic.ReplicationFactor)
+	d.Set("partitions", kafkaTopic.Partitions)
+
+	log.Printf("[INFO] Reading the config of topic %s.", topic)
+	kafkaTopicConfig, err := client.ReadKafkaTopicConfig(cluster_id, topic)
+	if err != nil {
+		return fmt.Errorf("[Error] Error reading Kafka topic %s's config: %w", topic, err)
+	}
+
+	config := make(map[string]interface{}, 0)
+	configList := make([]map[string]interface{}, 0)
+	err = mapstructure.Decode(kafkaTopicConfig.Config, &config)
+	if err != nil {
+		return fmt.Errorf("[Error] Error decoding the fetched kafka topic %s's config to a map: %w", topic, err)
+	}
+	configList = append(configList, config)
+	err = d.Set("config", configList)
+	if err != nil {
+		return fmt.Errorf("[Error] Error setting the kafka topic %s's config to Terraform state: %w", topic, err)
+	}
+
+	log.Printf("[INFO] Successfully fetch the config of topic %s.", topic)
+	return nil
+}
+
+// There are two use cases of resourceKafkaTopicUpdate:
+// 1. Providing all the possible configs in resource, and modify the ones you want to update. Then keeping all the
+// configs in the resource.
+// 2. Providing some configs (not all) you want to update, and then keep only these configs in the resource.
+// customDiffSuppressFunc is used to handle case 2, it makes sure that Terraform won't detect a change if some configs
+// are not provided in the resource.
+func resourceKafkaTopicUpdate(d *schema.ResourceData, meta interface{}) error {
+	cluster_id := d.Get("cluster_id").(string)
+	topic := d.Get("topic").(string)
+	log.Printf("[INFO] Updating the config of Kafka topic %s in %s.", topic, cluster_id)
+
+	// This is for when only replication_factor or partitions are changed, but we don't support changing them currently
+	if !d.HasChange("config") {
+		log.Printf("[INFO] Currently we only support updating topic's config. There are no changes in topic %s's config.", topic)
+		return nil
+	}
+	client := meta.(*Config).Client
+
+	changedConfigMap, err := getChangedConfigMap(d)
```

Comment:
I'm still not sure if this getChangedConfigMap method is necessary. Does the update API require all configuration properties to be provided? What if only the updated properties are provided?

Reply:
The purpose of this func is to give only the configs that are updated. Rather than giving all the configuration properties.